### PR TITLE
add default for HypervStrictCheck featureGate

### PR DESCRIPTION
**Pull request by kwiesmueller**
I'm not sure if this is the right way to do this. This will set the `HypervStrictCheck` FeatureGate to true if it (or any feature gates at all) are not provided.
It can be set to false by specifying it in the HCO config.

**Release note**:

```release-note
The kubevirt `HypervStrictCheck` FeatureGate will default to true.
```

/cc @ksimon1 @omeryahud

### CODE REVIEW

**Comment by kwiesmueller** [comment]
This is probably missing code generation etc.
Just want to clarify the approach first.

**Comment by omeryahud** [comment]
The direction looks good to me
@orenc1 @tiraboschi  Can you take a look please?

**Comment by tiraboschi** [comment]
can you please also run `make build-manifests` to regenerate the deploy manifests?

File: pkg/apis/hco/v1beta1/hyperconverged_types.go [hunk 1]
```diff
@@ -72,6 +72,11 @@ type HyperConvergedFeatureGates struct {
 	// Support migration for VMs with host-model CPU mode
 	// +optional
 	WithHostModelCPU *bool `json:"withHostModelCPU,omitempty"`
+
+	// Enable HyperV strict host checking for HyperV enlightenments
+	// Defaults to true, even when HyperConvergedFeatureGates is empty
```
Comment (nunnatsa) [inline pkg/apis/hco/v1beta1/hyperconverged_types.go:77]: I find it confusing. What do you think about setting this during reconciling? Somethink like this:

```golang
	if req.Instance.Spec.FeatureGates == nil {
		trueVal := true
		req.Instance.Spec.FeatureGates = &hcov1beta1.HyperConvergedFeatureGates{
			WithHostModelCPU: &trueVal,
		}
	}
```

File: pkg/controller/operands/kubevirt_test.go [hunk 2]
```diff
@@ -1385,21 +1386,29 @@ var _ = Describe("KubeVirt Operand", func() {
 
 		enabled := true
 		disabled := false
+		numOfDefaultGates := 1
```
Comment (nunnatsa) [inline pkg/controller/operands/kubevirt_test.go:1389]: Should be const, rather than var

**Review by nunnatsa — COMMENTED** [review]
Added online questions about how to enable features by default. @tiraboschi - what do you think?

File: pkg/apis/hco/v1beta1/hyperconverged_types.go (hunk 1 above)
Comment (tiraboschi) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:77]: proper defaulting on read?
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting

Comment (kwiesmueller) [reply pkg/apis/hco/v1beta1/hyperconverged_types.go:77]: Yeah, I think this should be the way to go: https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html

File: deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/kubevirt-hyperconverged-operator.v1.4.0.clusterserviceversion.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ metadata:
     categories: OpenShift Optional
     certified: "false"
     containerImage: quay.io/kubevirt/hyperconverged-cluster-operator@sha256:a137fd1ee55ae8806cbdd66912c7931c7f567315f0a927b3a12090112e96edb1
-    createdAt: "2021-01-28 18:56:06"
+    createdAt: "2021-01-31 03:41:38"
```
Comment (nunnatsa) [inline deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/kubevirt-hyperconverged-operator.v1.4.0.clusterserviceversion.yaml:12]: Please undo this change if it's the only change in this file. It just cause conflicts.

**Review by nunnatsa — CHANGES_REQUESTED** [review]
Please implement as in #1109 and #1113

**Review by nunnatsa — CHANGES_REQUESTED** [review]
Missing handling in `filterOutDisabledFeatureGates`.

Missing unit test for `IsHypervStrictCheckEnabled`.

**Comment by kwiesmueller** [comment]
Thanks for spotting those @nunnatsa !

**Review by nunnatsa — APPROVED** [review]
/approve

**Comment by nunnatsa** [comment]
I approve it even though it caused a new sonarCloud warning abut the complecity of `filterOutDisabledFeatureGates`...

**Comment by kwiesmueller** [comment]
@nunnatsa have a look at the new commit. Does this change make sense as a cleanup?
I'm not sure if I like the map much more than the switch but it should reduce complexity.

File: pkg/controller/operands/kubevirt.go [hunk 4]
```diff
@@ -294,35 +295,36 @@ func (h *kvConfigHooks) addEnabledFeatureGates(req *common.HcoRequest, foundFgSp
...
 		resultFg = append(resultFg, fg)
 	}
 	return fgChanged, resultFg
 }
 
+func isFeatureGateMissingFrom(checks featureGateChecks, featureGate string) bool {
+	if check, isKnown := checks[featureGate]; isKnown {
+		return !check()
+	}
+	return false
```
Comment (nunnatsa) [inline pkg/controller/operands/kubevirt.go:313]: I think it should be `true` here. If we get here, the FG **is** missing. Do you have a test to cover this (It's a bit confusing).

File: pkg/controller/operands/kubevirt.go [hunk 5]
```diff
@@ -294,35 +295,36 @@ func (h *kvConfigHooks) addEnabledFeatureGates(req *common.HcoRequest, foundFgSp
...
+}
+
+type featureGateChecks map[string]func() bool
+
+func getFeatureGateChecks(featureGates *hcov1beta1.HyperConvergedFeatureGates) featureGateChecks {
```
Comment (nunnatsa) [inline pkg/controller/operands/kubevirt.go:318]: I don't have strong opinion about this solution. I'm not sure it's not "too much". 

Regardless, maybe we can add a check for `featureGates == nil` before creating the map.

File: pkg/controller/operands/kubevirt.go [hunk 6]
```diff
@@ -528,5 +530,9 @@ func getKvFeatureGateList(fgs *hcov1beta1.HyperConvergedFeatureGates) []string {
 		res = append(res, SRIOVLiveMigrationGate)
 	}
 
+	if fgs.IsHypervStrictCheckEnabled() {
```
Comment (nunnatsa) [inline pkg/controller/operands/kubevirt.go:533]: Maybe we can use the `filterOutDisabledFeatureGates` method here as well?

**Review by nunnatsa — CHANGES_REQUESTED** [review]
Added 3 commnets. 2 of them may be ignored, but this one - https://github.com/kubevirt/hyperconverged-cluster-operator/pull/1094/files#r572610672 - I think it's a bug. Please re-check.

File: pkg/controller/operands/kubevirt.go (hunk 4 above)
Comment (kwiesmueller) [reply pkg/controller/operands/kubevirt.go:313]: I agree, switched it over while testing.
Will add a test.

File: pkg/controller/operands/kubevirt.go (hunk 5 above)
Comment (kwiesmueller) [reply pkg/controller/operands/kubevirt.go:318]: It will cause more allocations than the switch. Not sure if it's a concern though.

Comment (nunnatsa) [reply pkg/controller/operands/kubevirt.go:318]: Yes, and it called twice each time.

File: pkg/controller/operands/kubevirt_test.go [hunk 7]
```diff
@@ -1569,15 +1582,107 @@ var _ = Describe("KubeVirt Operand", func() {
...
+			}
+
+			changed, result := filterOutDisabledFeatureGates(
+				requestWithFeatureGates(fgs),
+				allGates,
+				make([]string, 0, len(allGates)))
+			Expect(changed).To(BeFalse())
+			Expect(result).To(Equal(allGates))
+		})
+		It("should remove gates that don't have a matching field", func() {
```
Comment (kwiesmueller) [inline pkg/controller/operands/kubevirt_test.go:1665]: @nunnatsa I'm not sure if that assumption is correct.
If my brain is not playing tricks on me the previous code did not remove gates that don't exist while it should have imo.
What do you think?

**Comment by kwiesmueller** [comment]
@nunnatsa could you please take another look?

**Comment by nunnatsa** [comment]
> @nunnatsa could you please take another look?

@kwiesmueller - sanity test failed. Details are here: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/1094/checks?check_run_id=1865547440

```
• Failure [0.000 seconds]
HyperconvergedController Reconcile HyperConverged Update Conflict Error [It] Should requeue in case of update conflict 
/home/runner/work/hyperconverged-cluster-operator/hyperconverged-cluster-operator/pkg/controller/hyperconverged/hyperconverged_controller_test.go:1457

  Expected
      <nil>: nil
  not to be nil

  /home/runner/work/hyperconverged-cluster-operator/hyperconverged-cluster-operator/pkg/controller/hyperconverged/hyperconverged_controller_test.go:1475
```

That usually mean that an unexpected error is returned somewhere in `doReconcile`

**Comment by nunnatsa** [comment]
Both `hack/ginkgo.sh pkg/controller/hyperconverged/...` and `hack/ginkgo.sh pkg/controller/operands/...` are failing

File: pkg/controller/operands/kubevirt_test.go (hunk 7 above)
Comment (kwiesmueller) [reply pkg/controller/operands/kubevirt_test.go]: I removed this test for now as it collided with existing behavior.

**Comment by kwiesmueller** [comment]
@nunnatsa the problem should be fixed.
It was a mix of the nil check and the behavior change I discussed in the other comment.

**Comment by nunnatsa** [comment]
/override coverage/coveralls - not less tests, but some refactoring cause changs in the coverage precentage.

/lgtm
/approve
